src/Accelerators/NNPA/NNPAAccelerator.cpp

```diff
@@ -25,15 +25,19 @@ extern llvm::cl::OptionCategory OMDLCPassOptions;
 extern llvm::cl::opt<DLCEmissionTargetType> dlcEmissionTarget;
 extern llvm::cl::list<std::string> execNodesOnCpu;
 
+mlir::NNPAAccelerator* pnnpa;
+
+void createNNPA() {
```

Comment:
[Suggestion]: Propose to use a factory method rather than a global function.

In `Accelerator.hpp`:
```
class Accelerator {
public:
  /// Kinds of accelerators.
  enum class Kind {
    NNPA, // IBM Telum coprocessor
  };

  virtual ~Accelerator();

  /// Create a new accelerator based on \p kind, initialize it, and add it to
  /// the list of available accelerators.
  static void create(Accelerator::Kind kind,
      mlir::OwningOpRef<mlir::ModuleOp> &module, mlir::MLIRContext &context,
      mlir::PassManager &pm, onnx_mlir::EmissionTargetType emissionTarget);
...
```

In `Accelerator.cpp`:
```
void Accelerator::create(Accelerator::Kind kind,
    mlir::OwningOpRef<mlir::ModuleOp> &module, mlir::MLIRContext &context,
    mlir::PassManager &pm, onnx_mlir::EmissionTargetType emissionTarget) {
  Accelerator *accel = nullptr;
  switch (kind) {
  case Kind::NNPA:
    accel = new mlir::NNPAAccelerator(); 
    if (!accel)
      llvm::errs() << "NNPA accelerator not supported\n";
    break;
  }
  assert(accel && "Accelerator not initialized correctly");

  // Initialize the new accelerator and add it to the list of available ones.
  accel->prepare(module, context, pm, emissionTarget);
  accelerators.insert(accel);
}
```

Reply:
> [Suggestion]: Propose to use a factory method rather than a global function.
> 
> In `Accelerator.hpp`:
> 
> ```
> class Accelerator {
> public:
>   /// Kinds of accelerators.
>   enum class Kind {
>     NNPA, // IBM Telum coprocessor
>   };
> 
>   virtual ~Accelerator();
> 
>   /// Create a new accelerator based on \p kind, initialize it, and add it to
>   /// the list of available accelerators.
>   static void create(Accelerator::Kind kind,
>       mlir::OwningOpRef<mlir::ModuleOp> &module, mlir::MLIRContext &context,
>       mlir::PassManager &pm, onnx_mlir::EmissionTargetType emissionTarget);
> ...
> ```
> 
> In `Accelerator.cpp`:
> 
> ```
> void Accelerator::create(Accelerator::Kind kind,
>     mlir::OwningOpRef<mlir::ModuleOp> &module, mlir::MLIRContext &context,
>     mlir::PassManager &pm, onnx_mlir::EmissionTargetType emissionTarget) {
>   Accelerator *accel = nullptr;
>   switch (kind) {
>   case Kind::NNPA:
>     accel = new mlir::NNPAAccelerator(); 
>     if (!accel)
>       llvm::errs() << "NNPA accelerator not supported\n";
>     break;
>   }
>   assert(accel && "Accelerator not initialized correctly");
> 
>   // Initialize the new accelerator and add it to the list of available ones.
>   accel->prepare(module, context, pm, emissionTarget);
>   accelerators.insert(accel);
> }
> ```
So, this is what I want to avoid ... this requires modifying the Accelerator class to add a new accelerator. I'm trying to completely contain knowledge about each accelerator inside its own subdirectory.

Reply:
OK, I see.  Using a factory method in the base class is pretty idiomatic but it does have the side effect of injecting some knowledge about the specific accelerators in the base Accelerator base class.  On way to limit such knowledge is by encapsulating it into a function:

```
void Accelerator::create(Accelerator::Kind kind,
    mlir::OwningOpRef<mlir::ModuleOp> &module, mlir::MLIRContext &context,
    mlir::PassManager &pm, onnx_mlir::EmissionTargetType emissionTarget) {
  Accelerator *accel = nullptr;
  switch (kind) {
  case Kind::ID:                                                               
    accel = nnpa::NNPAAccelerator::getInstance();                
    if (!accel)                                                                
      llvm::errs() << #ID " accelerator not supported\n";       
   }
```

where `getInstance()` is defined as a weak definition in `Accelerator.cpp` (returns nullptr) and defined to be a strong definition in `NNPAAccelerator.cpp`.  When no accelerator is built the linker uses the weak def., but when an accelerator is built the linker overwrites the weak definition with the strong one. 

In any case, I am not dead set on a particular solution. Limiting the need for the base class to "know" about derived classes is what is important.

Reply:
Also,  using __attribute__((weak))  is not ideal either because is not standard C++. Compilers I am familiar with (IBM XL, clang, gcc) all supports the attribute extension, and are source compatible... however looks like MSVC does not accept it (it has  __declspec(selectany) which does not have the exact semantics we need because it instructs the linker to select any def. available, non necessarily the strong definition).